### PR TITLE
Implement log in with devise_token_auth

### PR DESCRIPTION
**Pull request by aleochoam**
### Summary
- Added the gem devise_token_auth
- Implemented log in with devise_token_auth
- Integrate devise_token_auth to use the devise User legacy model
- Disabled routes for devise and add the devise_token_auth ones

### Trello Card

https://trello.com/c/SU3IUTP0/24-implementar-autenticaci%C3%B3n

### Screenshots

image

### CODE REVIEW

File: db/migrate/20180921151245_devise_create_users.rb [hunk 1]
```diff
@@ -4,12 +4,12 @@ class DeviseCreateUsers < ActiveRecord::Migration[5.1]
...
       ## Database authenticatable
-      t.string :email,              null: false, default: ""
-      t.string :encrypted_password, null: false, default: ""
+      t.string :email,              null: false, default: ''
+      t.string :encrypted_password, null: false, default: ''
 
       ## Recoverable
-      t.string   :reset_password_token
-      t.datetime :reset_password_sent_at
+      # t.string   :reset_password_token
```
Comment (Gmarn) [inline db/migrate/20180921151245_devise_create_users.rb]: Hummm, no, no and no. :)
Please never to any modification on migration. This could cause conflict between branch, transform a deploy in nightmare or complecate work of you collegue. If you need to change something on you database schema the only way (From a PR to another) is to create a new migration.

File: db/migrate/20180921151245_devise_create_users.rb [hunk 2]
```diff
@@ -4,12 +4,12 @@ class DeviseCreateUsers < ActiveRecord::Migration[5.1]
...
+      # t.datetime :reset_password_sent_at
```
Comment (Gmarn) [inline db/migrate/20180921151245_devise_create_users.rb]: same

File: db/migrate/20180921151245_devise_create_users.rb [hunk 3]
```diff
@@ -36,8 +36,8 @@ def change
       t.timestamps null: false
     end
 
-    add_index :users, :email,                unique: true
-    add_index :users, :reset_password_token, unique: true
+    add_index :users, :email, unique: true
+    # add_index :users, :reset_password_token, unique: true
```
Comment (Gmarn) [inline db/migrate/20180921151245_devise_create_users.rb]: same

File: db/migrate/20180921152413_add_fields_to_user.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 class AddFieldsToUser < ActiveRecord::Migration[5.1]
   def change
-    add_column :users, :first_name, :string
-    add_column :users, :last_name, :string
+    add_column :users, :first_name, :string, null: false
```
Comment (Gmarn) [inline db/migrate/20180921152413_add_fields_to_user.rb]: same

File: db/migrate/20180921152413_add_fields_to_user.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
...
+    add_column :users, :last_name, :string, null: false
```
Comment (Gmarn) [inline db/migrate/20180921152413_add_fields_to_user.rb]: same

File: db/migrate/20180921152413_add_fields_to_user.rb (hunk 4 above)
Comment (Gmarn) [reply db/migrate/20180921152413_add_fields_to_user.rb]: moreover here you should keep `first_name` as required

File: db/migrate/20180921152413_add_fields_to_user.rb (hunk 5 above)
Comment (Gmarn) [reply db/migrate/20180921152413_add_fields_to_user.rb]: Moreover here you should keep `last_name` as required

File: db/migrate/20180921153345_alter_users_model.rb [hunk 6]
```diff
@@ -1,6 +0,0 @@
-class AlterUsersModel < ActiveRecord::Migration[5.1]
```
Comment (Gmarn) [inline db/migrate/20180921153345_alter_users_model.rb]: Same

File: db/migrate/20180921151245_devise_create_users.rb [hunk 7]
```diff
@@ -4,12 +4,12 @@ class DeviseCreateUsers < ActiveRecord::Migration[5.1]
   def change
     create_table :users do |t|
       ## Database authenticatable
-      t.string :email,              null: false, default: ""
-      t.string :encrypted_password, null: false, default: ""
+      t.string :email,              null: false, default: ''
```
Comment (Gmarn) [inline db/migrate/20180921151245_devise_create_users.rb]: Let it as it was

File: db/migrate/20180921151245_devise_create_users.rb [hunk 8]
```diff
@@ -4,12 +4,12 @@ class DeviseCreateUsers < ActiveRecord::Migration[5.1]
...
+      t.string :encrypted_password, null: false, default: ''
```
Comment (Gmarn) [inline db/migrate/20180921151245_devise_create_users.rb]: Let it as it was

File: db/migrate/20180921151245_devise_create_users.rb [hunk 9]
```diff
@@ -36,8 +36,8 @@ def change
       t.timestamps null: false
     end
 
-    add_index :users, :email,                unique: true
-    add_index :users, :reset_password_token, unique: true
+    add_index :users, :email, unique: true
```
Comment (Gmarn) [inline db/migrate/20180921151245_devise_create_users.rb]: Let it as it was

File: Gemfile [hunk 10]
```diff
@@ -54,4 +54,6 @@ end
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
 gem 'devise'
+gem 'devise_token_auth'
+gem 'omniauth'
```
Comment (Gmarn) [inline Gemfile]: Why did you add this gem?

File: Gemfile [hunk 11]
```diff
@@ -54,4 +54,6 @@ end
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
 gem 'devise'
+gem 'devise_token_auth'
```
Comment (Gmarn) [inline Gemfile]: It's a better convention to specify the gem version when you add a new gem.

File: Gemfile (hunk 10 above)
Comment (aleochoam) [reply Gemfile]: omniauth is required by devise_token_auth to work properly even if it does not implement omniauth functions

File: Gemfile [hunk 12]
```diff
@@ -53,5 +53,7 @@ end
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
-gem 'devise'
+gem 'devise', '~> 4.5'
+gem 'devise_token_auth', '~> 1.0.0.rc2'
+gem 'omniauth', '~> 1.8.1'
```
Comment (Gmarn) [inline Gemfile]: I do not understand why you added this line?

Comment (aleochoam) [reply Gemfile]: Beacuse omniauth is required by devise_token_auth to work properly. Without it, when you start the server, it crashes because this gem is missing.

https://github.com/lynndylanhurley/devise_token_auth/issues/118
https://github.com/lynndylanhurley/devise_token_auth/issues/60
